test-utils/src/types.rs

```diff
@@ -72,3 +78,168 @@ impl WebSocketTestClient {
+        self.local_addr
+    }
+
+    pub async fn close(&mut self) {
+        self.exit.send(()).await.unwrap();
+    }
+}
+
+async fn server_backend(
+    listener: async_std::net::TcpListener,
+    mut exit: Receiver<()>,
+    mode: ServerMode,
+) {
+    let mut connections = Vec::new();
+
+    loop {
+        let conn_fut = listener.accept().fuse();
+        let exit_fut = exit.next();
+        futures::pin_mut!(exit_fut, conn_fut);
+
+        futures::select! {
+            exit = exit_fut => break,
+            conn = conn_fut => {
+                if let Ok((stream, _)) = conn {
+                    let (tx, rx) = mpsc::channel::<()>(4);
+                    let handle = tokio::spawn(connection_task(stream, mode.clone(), rx));
+                    connections.push((handle, tx));
+                }
+            }
+        }
+    }
+
+    // close connections
+    for (handle, mut exit) in connections {
+        // If the actual connection was never established i.e., returned early
+        // It will most likely be caught on the client-side but just to explicit.
+        exit.send(()).await.expect("WebSocket connection was never established");
+        handle.await.unwrap();
+    }
+}
+
+async fn connection_task(
+    socket: async_std::net::TcpStream,
+    mode: ServerMode,
+    mut exit: Receiver<()>,
+) {
+    let mut server = Server::new(socket);
+
+    let websocket_key = match server.receive_request().await {
+        Ok(req) => req.into_key(),
+        Err(_) => return,
+    };
+
+    let accept = server
+        .send_response(&Response::Accept {
+            key: &websocket_key,
+            protocol: None,
+        })
+        .await;
+
+    if accept.is_err() {
+        return;
+    }
+
+    let (mut sender, receiver) = server.into_builder().finish();
+
+    let ws_stream = stream::unfold(receiver, move |mut receiver| async {
+        let mut buf = Vec::new();
+        let ret = match receiver.receive_data(&mut buf).await {
+            Ok(_) => Ok(buf),
+            Err(err) => Err(err),
+        };
+        Some((ret, receiver))
+    });
+    futures::pin_mut!(ws_stream);
+
+    loop {
+        let next_ws = ws_stream.next().fuse();
+        let next_exit = exit.next().fuse();
+        let time_out = tokio::time::delay_for(Duration::from_secs(1)).fuse();
+        futures::pin_mut!(time_out, next_exit, next_ws);
+
+        futures::select! {
+            subscription = time_out => {
+                if let ServerMode::Subscription((_, r)) = &mode {
+                    if let Err(e) = sender.send_text(&r).await {
+                        log::warn!("send response to subscription: {:?}", e);
+                    }
+                }
+            }
+            ws = next_ws => {
+                // Got a request on the connection but don't care about the contents.
+                // Just send out the pre-configured hardcoded responses.
+                if let Some(Ok(_)) = ws {
+                    match &mode {
```

Comment:
this matching is dumb in loop because `mode` is known before the loop is entered but..... I didn't want to have more ugly if's or refactor to two different functions.